src/query/dblink_scan.c

```diff
@@ -853,6 +853,11 @@ dblink_scan_next (DBLINK_SCAN_INFO * scan_info, val_list_node * val_list)
 
 	  if (db_value_coerce (&cci_value, valptrp->val, &dom) != DOMAIN_COMPATIBLE)
 	    {
+	      /*
+	         in repeated error case, in order to display correct error message,
+	         we need to reinitialize the domain of valptrp->val
+	       */
+	      db_value_domain_init (valptrp->val, (DB_TYPE) dom.type->id, dom.precision, dom.scale);
```

Comment:
To preserve domain of `valptrp->val`, you can use `tp_value_cast_preserve_domain ()` instead of `db_value_coerce()`. In many places using `db_value_coerce()`, the destination value is temp or new allocated value.

Reply:
The function **tp_value_cast_preserve_domain** does not set the error.

Reply:
the error can be set by `tp_domain_status_er_set()`. object domain is already called above in this code. (`tp_domain_init()`).

Reply:
Also, for `VARCHAR` type, `codeset` and `collation` is not restored too by using the saved `TP_DOMAIN dom` above.

Reply:
However, **tp_value_cast_preserve_domain** does not call the tp_value_cast_internal with "**TP_FORCE_COERCION**".
whereas **db_value_coerce** calls the **tp_value_cast_internal** with "**TP_FORCE_COERCION**".

Reply:
Ok, I fixed it with **tp_value_cast_preserved_domain**